investos/portfolio/risk_model/factor_risk.py

```diff
@@ -1,23 +1,351 @@
+        Generate idio returns
+        """
+        print("\n generating idiosyncratic returns...")
+        # Merge the DataFrames on 'date'
+        self._df_idio = pd.merge(
+            self._df_loadings,
+            self._df_factor_returns.reset_index(),
+            on="datetime",
+            suffixes=("", "_factor_returns"),
+        )
+
+        # Multiplying matching columns
+        for col in self._fos_risk_factor_uuids + self._fos_industry_factors:
+            self._df_idio[f"calc_f_r_{col}"] = (
+                self._df_idio[col] * self._df_idio[f"returns_{col}"]
+            )
+
+        # Dropping the extra columns
+        self._df_idio = (
+            self._df_idio.drop(columns=self._fos_risk_factor_uuids)
+            .drop(columns=self._fos_industry_factors)
+            .drop(columns=self.cols_with_return)
+        )  # drop factor exposure columns
+        self._df_idio["factor_return_1d"] = self._df_idio[
+            [
+                f"calc_f_r_{col}"
+                for col in (self._fos_risk_factor_uuids + self._fos_industry_factors)
+            ]
+        ].sum(axis=1)
+        self._df_idio["factor_return_1d_error"] = (
+            self._df_idio["factor_return_1d"] - self._df_idio["return_1d"]
+        )
+        self._df_idio = self._df_idio[
+            [
+                "datetime",
+                "id",
+                "return_1d",
+                "factor_return_1d",
+                "factor_return_1d_error",
+            ]
+        ]
+        self._df_idio_returns = self._df_idio[
+            ["datetime", "id", "factor_return_1d_error"]
+        ].rename({"factor_return_1d_error": "idio_return"})
+        return
+
+    def _generate_factor_returns(self):
+        """
+        Generate factor returns
+        """
+        print("\n generating factor returns...")
+        list_to_insert = [
+            [
+                k,
+                self._factor_returns[k]["r2"],
+                *self._factor_returns[k]["t-values"][:-1],
+                *self._factor_returns[k]["coefficients"][:-1],
+            ]
+            for k in self._factor_returns
+        ]  # iterative through the factor returns dict
+
+        # split the columns into return, t-value and r2
+        self.cols_with_return = [
+            "returns_" + col
+            for col in (self._fos_risk_factor_uuids + self._fos_industry_factors)
+        ]
+        self.cols_with_t_values = [
+            "t_values_" + col
+            for col in (self._fos_risk_factor_uuids + self._fos_industry_factors)
+        ]
+        self._df_factor_summary = pd.DataFrame(
+            list_to_insert,
+            columns=[
+                "datetime",
+                "r2",
+                *self.cols_with_t_values,
+                *self.cols_with_return,
+            ],
+        ).set_index("datetime")
+        self._df_factor_returns = self._df_factor_summary[self.cols_with_return]
+        return
+
+    def _apply_risk_model_adjustments(self):
```

Comment:
There should be a separate method for building covariance matrix pre-adjustments

Reply:
As an aside, it wouldn't make sense to calculate the covariance matrix in a method called "_apply_risk_model_adjustments"